### PR TITLE
Get identity from certificate when using mTLS

### PR DESCRIPTION
**Pull request by nicoldr**
Signed-off-by: Nicol Draghici <idraghic@cisco.com>

**What type of PR is this?**
bug

**Which issue does this PR fix**:
resolves: #614

**What does this PR do / Why do we need it**:
Username unknown in case of certificates authentication. This PR solves the issue

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
Would refer to "username" as "identity". Change the language like so.

https://codeburst.io/mutual-tls-authentication-mtls-de-mystified-11fa2a52e9cf
^ for some context.

File: pkg/api/authz.go [hunk 1]
```diff
@@ -192,18 +192,32 @@ func AuthzHandler(ctlr *Controller) mux.MiddlewareFunc {
...
 				username, _, err = getUsernamePasswordBasicAuth(request)
-
-				if err != nil {
+				// no TLS mutual auth
+				if request.TLS != nil && err != nil && request.TLS.VerifiedChains == nil {
 					authFail(response, ctlr.Config.HTTP.Realm, ctlr.Config.HTTP.Auth.FailDelay)
 				}
 			}
 
+			// still no identity, get it from TLS certs
```
Comment (peusebiu) [inline pkg/api/authz.go:205]: move this comment on line 207

**Review by peusebiu — COMMENTED** [review]
From my understanding we just need to try and take the username from certificates if there is no username from basic auth and leave the rest of the code as is.

**Comment by peusebiu** [comment]
> Would refer to "username" as "identity". Change the language like so.
> 
> https://codeburst.io/mutual-tls-authentication-mtls-de-mystified-11fa2a52e9cf ^ for some context.

but in our case username is the subject in authorization, I don't understand, what do you want to change?

**Comment by peusebiu** [comment]
The code is getting hard to grasp, there are lots of ifs, how do we want to treat this?

I would say username:password from basic auth should have bigger priority than mtls.
A) basic auth enabled and tls CACert
1. if username:password given then use that. (even if there are verified certs)
2. if no username:password given then try to get username(identity) from certificates.
3. if no username:password nor identity from certs then it's anonymous.
4. if bad username:password but verified certs the authFail().

B) basic auth disabled and tls CACert
1. get username(identity) from certificates.

**Review by andaaron — COMMENTED** [review]
@peusebiu 
Not sure about the "identity" change. Should we refer to the username from basic auth or the cn from cert auth, as "identify" in variable names in both cases?

File: pkg/api/authz.go [hunk 2]
```diff
@@ -192,18 +192,32 @@ func AuthzHandler(ctlr *Controller) mux.MiddlewareFunc {
 			var username string
 			var err error
 
-			/* To be implemented: verify client certs and get its username(subject DN)
-			if request.TLS.VerifiedChains != nil, then get subject DN
-			issue: https: //github.com/project-zot/zot/issues/614 */
-
+			// allow anonymous authz if no authn present and only default policies are present
+			username = ""
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/authz.go:196]: yeah, makes sense :D

File: pkg/api/authz.go [hunk 3]
```diff
@@ -192,18 +192,32 @@ func AuthzHandler(ctlr *Controller) mux.MiddlewareFunc {
...
 
+			// still no identity, get it from TLS certs
+			if request.TLS != nil {
+				verifiedChains := request.TLS.VerifiedChains
+				if username == "" && verifiedChains != nil &&
+					len(verifiedChains) > 0 && len(verifiedChains[0]) > 0 {
+					for _, cert := range request.TLS.PeerCertificates {
+						username = cert.Subject.CommonName
+					}
+				}
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/authz.go:213]: agree

**Comment by rchincha** [comment]
@nicoldr lgtm, but can you pls resolve comments so that we haven't missed anything.

**Comment by nicoldr** [comment]
> @nicoldr lgtm, but can you pls resolve comments so that we haven't missed anything.

All conversations are resolved and PR has all checks now

**Review by andaaron — APPROVED** [review]
lgtm

**Pull request merged**